### PR TITLE
pubsub emulator rule moved to a separate project

### LINKED ISSUE #2351: Make PubSub emulator rule public
Creating new module for the future emulators support.

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #2351

cc/ @elefeint

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -45,6 +45,7 @@
 		<module>spring-cloud-gcp-bigquery</module>
 		<module>spring-cloud-gcp-security-firebase</module>
 		<module>spring-cloud-gcp-secretmanager</module>
+		<module>spring-cloud-gcp-test-support</module>
```
Comment (meltsufin) [inline pom.xml:48]: Let's rename to `spring-cloud-gcp-test`.

File: spring-cloud-gcp-test/pom.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    </parent>
+
+
+    <artifactId>spring-cloud-gcp-test</artifactId>
+    <name>Spring Cloud GCP Test Support</name>
+    <description>Provides tools for testing Spring Cloud GCP projects</description>
+    <dependencies>
+		<dependency>
+			<groupId>junit</groupId>
+			<artifactId>junit</artifactId>
```
Comment (meltsufin) [inline spring-cloud-gcp-test/pom.xml:19]: `test` scope?

File: spring-cloud-gcp-test/pom.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+		</dependency>
+
+		<dependency>
+			<groupId>org.springframework</groupId>
+			<artifactId>spring-jcl</artifactId>
+			<scope>compile</scope>
```
Comment (meltsufin) [inline spring-cloud-gcp-test/pom.xml:25]: Compile scope is the default, you don't have to specify it.

File: spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java [hunk 4]
```diff
@@ -87,7 +86,7 @@ public PubSubEmulator() {
 	@Override
 	protected void before() throws IOException, InterruptedException {
 
-		assumeTrue("PubSubEmulator rule disabled. Please enable with -Dit.pubsub-emulator.", this.enableTests);
+		Assume.assumeTrue("PubSubEmulator rule disabled. Please enable with -Dit.pubsub-emulator.", this.enableTests);
```
Comment (meltsufin) [inline spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java:89]: I don't think that enabling it with `-Dit.pubsub-emulator` makes sense when you're offering it as a generic test utility. That flag was specific for our project.

File: spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java [hunk 5]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package org.springframework.cloud.gcp.stream.binder.pubsub;
+package org.springframework.cloud.gcp.test;
```
Comment (meltsufin) [inline spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java:17]:
```suggestion
package org.springframework.cloud.gcp.test.pubsub;
```

File: spring-cloud-gcp-test/pom.xml (hunk 2 above)
Comment (elefeint) [reply spring-cloud-gcp-test/pom.xml:19]: Not in this case -- the test project itself is meant to be imported as test scope, but the rule extends a JUnit class, so it needs to be compile, not test scope.

File: spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java (hunk 4 above)
Comment (elefeint) [reply spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java:89]: It's an interesting question then to test whether the rule initialization or a test's `@BeforeClass` will run first. We would not want to do emulator startup if the tests themselves are disabled.

If there is possibility of the rule running first, perhaps we can just document this flag well?

File: spring-cloud-gcp-test/pom.xml (hunk 2 above)
Comment (meltsufin) [reply spring-cloud-gcp-test/pom.xml:19]: I wonder if we should consider making these utilities independent of JUnit. People use different test frameworks.

File: spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java (hunk 4 above)
Comment (meltsufin) [reply spring-cloud-gcp-test/src/main/java/org/springframework/cloud/gcp/test/PubSubEmulator.java:89]: We can also try not to implement JUnit test rule.
That's what we ended up doing for Spring Cloud Function [here](https://github.com/spring-cloud/spring-cloud-function/blob/master/spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/test/java/org/springframework/cloud/function/adapter/gcp/integration/LocalServerTestSupport.java).

File: spring-cloud-gcp-test/pom.xml (hunk 2 above)
Comment (eddumelendez) [reply spring-cloud-gcp-test/pom.xml:19]: make it generic so people can reuse it sounds good. Can you take into account that may some people will be able to contribute creating rules for junit4 or extension for junit5 specifically? Those can be issues with `ideal for contribution` or `help wanted`. WDYT?

Comment (dmitry-s) [reply spring-cloud-gcp-test/pom.xml:19]: That a good idea! We should do that.

**Comment by dmitry-s** [comment]
superseded by #2357

**Pull request closed without merging**